adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Flows/AcquireTokenNonInteractiveHandler.cs

```diff
@@ -126,12 +130,52 @@ public AcquireTokenNonInteractiveHandler(RequestData requestData, UserAssertion
                         throw new AdalException(AdalError.MissingFederationMetadataUrl);
                     }
-                    
-                    WsTrustAddress wsTrustAddress = await MexParser.FetchWsTrustAddressFromMexAsync(userRealmResponse.FederationMetadataUrl, this.userCredential.UserAuthType, RequestContext).ConfigureAwait(false);
+
+                    WsTrustAddress wsTrustAddress = null;
+                    try
+                    {
+                        wsTrustAddress = await MexParser.FetchWsTrustAddressFromMexAsync(userRealmResponse.FederationMetadataUrl, this.userCredential.UserAuthType, RequestContext).ConfigureAwait(false);
+                        if (wsTrustAddress == null)
+                        {
+                            if (this.userCredential.UserAuthType == Identity.Core.UserAuthType.IntegratedAuth)
+                            {
+                                throw new AdalException(AdalError.IntegratedAuthFailed, new AdalException(AdalError.WsTrustEndpointNotFoundInMetadataDocument));
+                            }
+                            else
+                            {
+                                throw new AdalException(AdalError.WsTrustEndpointNotFoundInMetadataDocument);
+                            }
+                        }
+                    }
+                    catch (System.Xml.XmlException ex)
+                    {
+                        throw new AdalException(AdalError.ParsingWsMetadataExchangeFailed, ex);
+                    }
+                    catch (Identity.Client.MsalException ex)
+                    {
+                        throw new AdalException(AdalError.AccessingWsMetadataExchangeFailed, ex);
+                    }
                     RequestContext.Logger.InfoPii(string.Format(CultureInfo.CurrentCulture, " WS-Trust endpoint '{0}' fetched from MEX at '{1}'",
                             wsTrustAddress.Uri, userRealmResponse.FederationMetadataUrl));
 
-                    WsTrustResponse wsTrustResponse = await WsTrustRequest.SendRequestAsync(wsTrustAddress, this.userCredential, RequestContext, userRealmResponse.CloudAudienceUrn).ConfigureAwait(false);
+                    WsTrustResponse wsTrustResponse;
+                    try
+                    {
+                        wsTrustResponse = await WsTrustRequest.SendRequestAsync(wsTrustAddress, this.userCredential._UserCredential, RequestContext, userRealmResponse.CloudAudienceUrn).ConfigureAwait(false);
+                    }
+                    catch (System.Xml.XmlException ex)
+                    {
+                        throw new AdalException(AdalError.ParsingWsTrustResponseFailed, ex);
+                    }
+                    catch (Identity.Client.MsalServiceException ex)
+                    {
+                        throw new AdalException(AdalError.ParsingWsTrustResponseFailed, ex.Message, ex);
+                    }
+                    if (wsTrustResponse == null)
```

Comment:
If you add a newline here, this would be easier to read, otherwise the if seems to be part of the try.

Reply:
Actually I did it on purpose for exactly the same reason. This extra `if(...) {throw ...}` is semantically part of the exception handling.